File: day03.js
/* https://adventofcode.com/2022/day/3 */

function find_repeat(sack) {
    compartment1 = sack.substring(0, sack.length / 2);
    compartment2 = sack.substring(sack.length / 2);
    for (char of compartment1) {
        if (compartment2.includes(char)) {
            return char
        }
    }
}


function score_priority(char) {
    ordering = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';
    return ordering.indexOf(char) + 1;
}


function sum_priorities(sacks) {
    total = 0;
    for (sack of sacks) {
        repeated = find_repeat(sack);
        priority = score_priority(repeated);
        total = total + priority;
    }
    return total
}


async function main() {
    /* Test with example */
    example = [
        'vJrwpWtwJgWrhcsFMMfFFhFp',
        'jqHRNqRjqzjGDLGLrsFMfFZSrLrFZsSL',
        'PmmdzqPrVvPwwTWBwg',
        'wMqvLMZHhHMvwLHjbvcjnnSBnvTQFn',
        'ttgJtRGJQctTZtZT',
        'CrZsJsPPZsGzwwsLwLmpwMDw'
    ];
    example_res = sum_priorities(example);
    if (example_res != 157) {
        console.log('Expected 157, got ' + example_res);
        throw new Error('Failed the example!');
    }
    const infile = new File([], 'day03.txt', {type: 'text/plain'});
    reader = new FileReader();
    reader.readAsText(infile);
    const intext = reader.result;
    rucksacks = intext.result.split('\n');
    result1 = sum_priorities(rucksacks);
    console.log(result1);
}


main()

File: day03.py
"""https://adventofcode.com/2022/day/3"""

from functools import reduce


def find_repeats(sacks):
    repeated = []
    for sack in sacks:
        halfway = len(sack) // 2
        compartment1 = set(sack[:halfway])
        compartment2 = set(sack[halfway:])
        repeated.extend(compartment1 & compartment2)
    return repeated


def total_priority(chars):
    ordering = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'
    return sum(ordering.index(char) + 1 for char in chars)


def sum_repeat_priority(sacks):
    """
    >>> rs = [
    ...     'vJrwpWtwJgWrhcsFMMfFFhFp',
    ...     'jqHRNqRjqzjGDLGLrsFMfFZSrLrFZsSL',
    ...     'PmmdzqPrVvPwwTWBwg',
    ...     'wMqvLMZHhHMvwLHjbvcjnnSBnvTQFn',
    ...     'ttgJtRGJQctTZtZT',
    ...     'CrZsJsPPZsGzwwsLwLmpwMDw',
    ... ]
    >>> sum_repeat_priority(rs)
    157
    """
    repeats = find_repeats(sacks)
    return total_priority(repeats)


def sum_trio_badges(sacks):
    """
    >>> rs = [
    ...     'vJrwpWtwJgWrhcsFMMfFFhFp',
    ...     'jqHRNqRjqzjGDLGLrsFMfFZSrLrFZsSL',
    ...     'PmmdzqPrVvPwwTWBwg',
    ...     'wMqvLMZHhHMvwLHjbvcjnnSBnvTQFn',
    ...     'ttgJtRGJQctTZtZT',
    ...     'CrZsJsPPZsGzwwsLwLmpwMDw',
    ... ]
    >>> sum_trio_badges(rs)
    70
    """
    sacks = list(sacks)
    badges = []
    while sacks:
        trio = [set(ss) for ss in sacks[-3:]]
        del sacks[-3:]
        common = reduce(lambda x, y: x & y, trio)
        assert len(common) == 1
        badges.append(common.pop())
    return total_priority(badges)


if __name__ == '__main__':
    import doctest
    doctest.testmod()
    with open('day03.txt') as infile:
        rucksacks = [line.strip() for line in infile.readlines()]
    # Part 1
    print(sum_repeat_priority(rucksacks))
    # Part 2
    print(sum_trio_badges(rucksacks))
